File: backend/src/modules/assessment/scoring.service.js
const DIFFICULTY_POINTS = { easy: 1, medium: 2, hard: 3 };

function calculateRawScore(answers, questions) {
  let rawScore = 0;
  let maxPossibleScore = 0;
  for (let i = 0; i < questions.length; i++) {
    const q = questions[i];
    const pts = DIFFICULTY_POINTS[q.difficulty] || 1;
    maxPossibleScore += pts;
    const answer = answers[q.id] ?? answers[i];
    const correctIndex = q.correctIndex ?? q.correct;
    if (answer === correctIndex) rawScore += pts;
  }
  return { rawScore, maxPossibleScore };
}

function normaliseScore(rawScore, maxPossibleScore) {
  if (maxPossibleScore <= 0) return 0;
  return Math.round((rawScore / maxPossibleScore) * 100);
}

function assignTier(score) {
  if (score >= 80) return 'Advanced';
  if (score >= 60) return 'Proficient';
  if (score >= 40) return 'Developing';
  return 'Beginner';
}

function identifyGaps(skillScores, benchmarks) {
  const gaps = [];
  for (const catId of Object.keys(benchmarks || {})) {
    const userScore = (skillScores || []).find(s => s.categoryId === catId)?.currentScore ?? 0;
    const bench = benchmarks[catId] ?? 60;
    const gap = bench - userScore;
    if (gap > 0) gaps.push({ categoryId: catId, gapPoints: gap, benchmark: bench, currentScore: userScore });
  }
  gaps.sort((a, b) => b.gapPoints - a.gapPoints);
  return gaps;
}

function generateRadarData(skillScores) {
  const cats = ['digital-literacy', 'communication', 'business-entrepreneurship', 'technical-ict', 'soft-skills-leadership'];
  return cats.map(catId => ((skillScores || []).find(s => s.categoryId === catId)?.currentScore ?? 0) / 100);
}

module.exports = { calculateRawScore, normaliseScore, assignTier, identifyGaps, generateRadarData, DIFFICULTY_POINTS };

File: lib/features/assessment/domain/services/scoring_service.dart
/// Pure scoring logic for assessments (mirrors backend scoring.service.js).
/// Used by assessment repository and tests.

class SkillScore {
  const SkillScore({required this.categoryId, required this.currentScore});
  final String categoryId;
  final int currentScore;
}

class GapItemResult {
  const GapItemResult({
    required this.categoryId,
    required this.gapPoints,
    this.benchmark,
    this.currentScore,
  });
  final String categoryId;
  final int gapPoints;
  final int? benchmark;
  final int? currentScore;
}

class ScoringService {
  ScoringService._();

  static const Map<String, int> difficultyPoints = {'easy': 1, 'medium': 2, 'hard': 3};

  /// 0 raw returns 0; maxRaw returns 100; mid returns correct value.
  static int normaliseScore(int rawScore, int maxPossibleScore) {
    if (maxPossibleScore <= 0) return 0;
    return (rawScore / maxPossibleScore * 100).round().clamp(0, 100);
  }

  /// assignTier: 0=Beginner, 39=Beginner, 40=Developing, 60=Proficient, 80=Advanced
  static String assignTier(int score) {
    if (score >= 80) return 'Advanced';
    if (score >= 60) return 'Proficient';
    if (score >= 40) return 'Developing';
    return 'Beginner';
  }

  /// Returns gaps sorted by severity (highest gap first); excludes no-gap categories.
  static List<GapItemResult> identifyGaps(
    List<SkillScore> skillScores,
    Map<String, int> benchmarks,
  ) {
    final gaps = <GapItemResult>[];
    for (final entry in benchmarks.entries) {
      final categoryId = entry.key;
      final bench = entry.value;
      final matching = skillScores.where((s) => s.categoryId == categoryId).toList();
      final userScore = matching.isEmpty ? 0 : matching.first.currentScore;
      final gap = bench - userScore;
      if (gap > 0) {
        gaps.add(GapItemResult(
          categoryId: categoryId,
          gapPoints: gap,
          benchmark: bench,
          currentScore: userScore,
        ));
      }
    }
    gaps.sort((a, b) => b.gapPoints.compareTo(a.gapPoints));
    return gaps;
  }
}
